internal/buildkitd/docker.provisioner.go

```diff
@@ -0,0 +1,253 @@
+package buildkitd
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"io/fs"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/docker/distribution/reference"
+	bkclient "github.com/moby/buildkit/client"
+	"github.com/rs/zerolog/log"
+	"go.dagger.io/dagger"
+)
+
+const (
+	dockerfileName = "Dockerfile.daggerd"
+)
+
+// ensure the docker CLI is available and properly set up (e.g. permissions to
+// communicate with the daemon, etc)
+func checkDocker(ctx context.Context) error {
+	cmd := exec.CommandContext(ctx, "docker", "info")
+	output, err := cmd.CombinedOutput()
+	if err != nil {
+		log.
+			Ctx(ctx).
+			Error().
+			Err(err).
+			Bytes("output", output).
+			Msg("failed to run docker")
+		return fmt.Errorf("%s%s", err, output)
+	}
+
+	return nil
+}
+
+type Docker struct {
+	host string
+}
+
+func (d Docker) GetHost() string {
+	return d.host
+}
+
+// create a copy of an embed directory
+func copyDir(src fs.FS, dst string) error {
+	return fs.WalkDir(src, ".", func(path string, d fs.DirEntry, err error) error {
+		if err != nil {
+			return err
+		}
+
+		if d.IsDir() {
+			return os.MkdirAll(filepath.Join(dst, path), 0755)
+		}
+
+		// #nosec
+		srcFile, err := src.Open(path)
+		if err != nil {
+			return err
+		}
+		defer srcFile.Close()
+
+		// #nosec
+		dstFile, err := os.Create(filepath.Join(dst, path))
+		if err != nil {
+			return err
+		}
+		defer dstFile.Close()
+
+		_, err = io.Copy(dstFile, srcFile)
+		return err
+	})
+}
+
+func (d Docker) buildDaggerd(ctx context.Context, version string) error {
+	dirPath, err := os.MkdirTemp("", "daggerd")
+	if err != nil {
+		return err
+	}
+
+	defer os.RemoveAll(dirPath)
+	if err := copyDir(dagger.SourceCode, dirPath); err != nil {
+		return err
+	}
+
+	fmt.Println("Building daggerd image...")
+
+	// #nosec
+	// Workaround to avoid:
+	// failed to solve with frontend dockerfile.v0: failed to read dockerfile: Dockerfile.daggerd: no such file or directory
+	// Manually create "Dockerfile" from "Dockerfile.daggerd"
+	cmd := exec.CommandContext(ctx,
+		"cp",
+		dirPath+"/"+dockerfileName,
+		dirPath+"/"+"Dockerfile",
+	)
+	output, err := cmd.CombinedOutput()
+	if err != nil {
+		return fmt.Errorf("cp error: %s\noutput:%s", err, output)
+	}
+
+	// #nosec
+	// move to build operation
+	cmd = exec.CommandContext(ctx,
```

Comment:
Saw the test failures in GHA, think you may just need to set the `DOCKER_BUILDKIT=1` env var when calling this.

Reply:
You were right. Thanks 🙏